rust/src/fvm/engine.rs

```diff
@@ -0,0 +1,239 @@
+    DefaultMachine as DefaultMachine2, MachineContext as MachineContext2,
+    MultiEngine as MultiEngine2, NetworkConfig as NetworkConfig2,
+};
+use fvm3::machine::{
+    DefaultMachine as DefaultMachine3, MachineContext, MultiEngine as MultiEngine3, NetworkConfig,
+};
+
+use fvm2::DefaultKernel as DefaultKernel2;
+use fvm3::DefaultKernel as DefaultKernel3;
+
+use fvm2::gas::PriceList as PriceList2;
+use fvm3::gas::PriceList as PriceList3;
+
+use fvm3_shared::{econ::TokenAmount as TokenAmount3, message::Message, version::NetworkVersion};
+
+use fvm2_shared::{
+    econ::TokenAmount as TokenAmount2, message::Message as Message2,
+    version::NetworkVersion as NetworkVersion2,
+};
+
+use fvm2_shared::address::Address as Address2;
+
+use fvm2_ipld_encoding::RawBytes as RawBytes2;
+use fvm3_ipld_encoding::RawBytes as RawBytes3;
+
+use super::blockstore::{CgoBlockstore, OverlayBlockstore};
+use super::externs::CgoExterns;
+use super::types::*;
+
+pub type CgoMachine3 = DefaultMachine3<OverlayBlockstore<CgoBlockstore>, CgoExterns>;
+pub type BaseExecutor3 = DefaultExecutor3<DefaultKernel3<DefaultCallManager3<CgoMachine3>>>;
+pub type CgoExecutor3 = ThreadedExecutor3<BaseExecutor3>;
+
+pub type CgoMachine2 = DefaultMachine2<OverlayBlockstore<CgoBlockstore>, CgoExterns>;
+pub type BaseExecutor2 = DefaultExecutor2<DefaultKernel2<DefaultCallManager2<CgoMachine2>>>;
+pub type CgoExecutor2 = ThreadedExecutor2<BaseExecutor2>;
+
+fn new_executor3(machine: CgoMachine3) -> CgoExecutor3 {
+    ThreadedExecutor3(BaseExecutor3::new(machine))
+}
+
+fn new_executor2(machine: CgoMachine2) -> CgoExecutor2 {
+    ThreadedExecutor2(BaseExecutor2::new(machine))
+}
+
+pub trait CgoExecutor {
+    fn execute_message(
+        &mut self,
+        msg: Message,
+        apply_kind: ApplyKind,
+        raw_length: usize,
+    ) -> anyhow::Result<ApplyRet>;
+
+    fn flush(&mut self) -> anyhow::Result<Cid>;
+}
+
+impl CgoExecutor for CgoExecutor3 {
+    fn execute_message(
+        &mut self,
+        msg: Message,
+        apply_kind: ApplyKind,
+        raw_length: usize,
+    ) -> anyhow::Result<ApplyRet> {
+        use fvm3::executor::Executor;
+        self.0.execute_message(msg, apply_kind, raw_length)
+    }
+
+    fn flush(&mut self) -> anyhow::Result<Cid> {
+        use fvm3::executor::Executor;
+        self.0.flush()
+    }
+}
+
+impl CgoExecutor for CgoExecutor2 {
+    fn execute_message(
+        &mut self,
+        msg: Message,
+        apply_kind: ApplyKind,
+        raw_length: usize,
+    ) -> anyhow::Result<ApplyRet> {
+        use fvm2::executor::Executor;
+        let res = self.0.execute_message(
+            Message2 {
+                version: msg.version,
+                from: Address2::from_bytes(&msg.from.to_bytes()).unwrap(),
+                to: Address2::from_bytes(&msg.to.to_bytes()).unwrap(),
+                sequence: msg.sequence,
+                value: unsafe { std::mem::transmute::<TokenAmount3, TokenAmount2>(msg.value) },
```

Comment:
can we avoid this? It won't really cost anything.

Reply:
Yeah, let's get rid of the transmutes. We won't notice if something changes.

Reply:
removed all the transmutes; this was painful, but much better than have a random (maybe silent) failure.

Obligatory comment about rust: it is insane that the layout of structs and stuff is not deterministic. This is the biggest advantage of C, you can rely on things being the same and laid out the same.

Reply:
Obligatory reply to rust comment :) If you want C semantics, you can annotate the structs with [`#[repr(C)]` ](https://doc.rust-lang.org/reference/type-layout.html#the-c-representation)`.

Reply:
I dont necessarily want C repr, i jist want the compiler to be sane and predictable damnit!

Side note: i have somewhat warmed to that language of yours, but i still get fits from all those little things.